### PR TITLE
[IT-2356] Setup secrets rotation

### PR DESCRIPTION
**Pull request by zaro0508**
Setup DB secrets on a rotation schedule for enhanced security.

### CODE REVIEW

File: templates/nextflow-aurora-mysql.yaml
Comment (BrunoGrandePhD) [inline templates/nextflow-aurora-mysql.yaml:107]: Doesn't this need to be paired with a Lambda function to update the database accordingly? For example, using the `HostedRotationLambda` property ([docs](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-secretsmanager-rotationschedule-hostedrotationlambda.html)). I've never configured secret rotation, so I might be completely off-base.

**Review by BrunoGrandePhD — CHANGES_REQUESTED** [review]

File: templates/nextflow-aurora-mysql.yaml
Comment (zaro0508) [reply templates/nextflow-aurora-mysql.yaml:107]: you are correct.  The docs says configuring `HostedRotationLambda` was not required so i was hoping that AWS would just figure it out for us.  I've tested and that is indeed not the case.

Comment (zaro0508) [reply templates/nextflow-aurora-mysql.yaml:110]: done, i have set to use the AWS built in `MySQLSingleUser` lambda and explicitly set the rotation schedule to 30 days.

**Comment by thomasyu888** [comment]
Thanks @zaro0508  for your PRs to nextflow-infra. I just wanted to acknowledge that we received these PRs and that we will be reviewing them soon.

We are currently testing the wave feature in Tower so wanted to hold off on these.

File: templates/nextflow-aurora-mysql.yaml
Comment (BrunoGrandePhD) [inline templates/nextflow-aurora-mysql.yaml:112]: I was getting a linting warning that this property depends on a resource with a condition (`Condition: UseSnapshotFalse`), so there's a risk that the referred resource doesn't exist. You might want to add `Condition: UseSnapshotFalse` to this resource as well.

**Review by BrunoGrandePhD — CHANGES_REQUESTED** [review]
I manually deployed this to `nextflow-dev` and tried rotating the secret. The Lambda function failed with the following error. As far as I can tell, the Lambda function doesn't have access to the KMS key that's used to encrypt the secret (_i.e._ the `AuroraKeyStack` nested stack in this file). 

```
[ERROR] ClientError: An error occurred (AccessDeniedException) when calling the GetSecretValue operation: Access to KMS is not allowed
Traceback (most recent call last):
  File "/var/task/lambda_function.py", line 72, in lambda_handler
    create_secret(service_client, arn, token)
  File "/var/task/lambda_function.py", line 108, in create_secret
    current_dict = get_secret_dict(service_client, arn, "AWSCURRENT")
  File "/var/task/lambda_function.py", line 416, in get_secret_dict
    secret = service_client.get_secret_value(SecretId=arn, VersionStage=stage)
  File "/var/runtime/botocore/client.py", line 391, in _api_call
    return self._make_api_call(operation_name, kwargs)
  File "/var/runtime/botocore/client.py", line 719, in _make_api_call
    raise error_class(parsed_response, operation_name)
```

File: templates/nextflow-aurora-mysql.yaml
Comment (zaro0508) [reply templates/nextflow-aurora-mysql.yaml:113]: done.

File: templates/nextflow-aurora-mysql.yaml
Comment (BrunoGrandePhD) [inline templates/nextflow-aurora-mysql.yaml:109]: I don't think you address the KMS permission issue that I described [here](https://github.com/Sage-Bionetworks-Workflows/nextflow-infra/pull/168#pullrequestreview-1370106587).

**Review by BrunoGrandePhD — CHANGES_REQUESTED** [review]

File: templates/nextflow-aurora-mysql.yaml
Comment (zaro0508) [reply templates/nextflow-aurora-mysql.yaml:109]: ahh yes, sorry about that.  I've updated and I think it should work now.  here are my updates..

* Update kms-key template to v0.6.83 which contains
an KeyArn[1] as an output and more expansive permission[2].

* Add master secrets to the rotation schedule.  The docs for
the MasterSecretKmsKeyArn[3] says that..
```
CloudFormation grants the execution role for the Lambda
rotation function Decrypt, DescribeKey, and GenerateDataKey
permission to the key in this property
```

[1] https://github.com/Sage-Bionetworks/aws-infra/commit/04c5b11c4907c76917e8a13b3c03779dccd0d1ca
[2] https://github.com/Sage-Bionetworks/aws-infra/commit/177f63228cdb5e6a2933b88b6049ba01a372c0ae
[3] https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-secretsmanager-rotationschedule-hostedrotationlambda.html#cfn-secretsmanager-rotationschedule-hostedrotationlambda-mastersecretarn

File: templates/nextflow-aurora-mysql.yaml
Comment (BrunoGrandePhD) [inline templates/nextflow-aurora-mysql.yaml:117]: When I was testing the latest commit, I ran into an error. I was able to overcome it by switching from `MasterSecretKmsKeyArn` to `KmsKeyArn`. It seems that the `MasterSecret*` parameters are only relevant with the alternating user setup. 

```suggestion
        KmsKeyArn: !Sub ${AuroraKeyStack.Outputs.KeyArn}
```

File: templates/nextflow-aurora-mysql.yaml
Comment (BrunoGrandePhD) [inline templates/nextflow-aurora-mysql.yaml:109]: I'm still unable to manually rotate the secret after deploying the stack in `nextflow-dev`. I run into the following error: 

```
Failed to rotate the secret "nextflow-aurora-mysql-MasterSecret". A previous rotation isn't complete. That rotation will be reattempted.
```

image

**Review by BrunoGrandePhD — CHANGES_REQUESTED** [review]

File: templates/nextflow-aurora-mysql.yaml
Comment (zaro0508) [reply templates/nextflow-aurora-mysql.yaml:109]: @BrunoGrandePhD would you like to give this a try? https://stackoverflow.com/a/50337928/1094247

i can also give it try if you prefer that i debug.

Comment (BrunoGrandePhD) [reply templates/nextflow-aurora-mysql.yaml:109]: I'll look into that tomorrow

Comment (BrunoGrandePhD) [reply templates/nextflow-aurora-mysql.yaml:109]: Indeed, there was a version stuck in `AWSPENDING`. I cleared that stage using the following command. 

```console
aws secretsmanager update-secret-version-stage --secret-id <arn> --version-stage AWSPENDING --remove-from-version-id <version-id>
```

After, I tried rotating the secret manually, and I got a green pop-up message instead of a red one. However, I don't think the Lambda succeeded. So, this might still need some work. I'll defer to Khai.

**Comment by zaro0508** [comment]
Amazon Aurora recently added integration with Secrets Manager to manage master user passwords for DB clusters[1].  This PR has been replaced with that implementation in PR #177 

[1] https://docs.aws.amazon.com/AmazonRDS/latest/AuroraUserGuide/rds-secrets-manager.html

File: templates/nextflow-aurora-mysql.yaml
Comment (zaro0508) [reply templates/nextflow-aurora-mysql.yaml:117]: Yup, i have confirmed that deployment only worked when removing Master secret with just the `KmsKeyArn`.  After deploying successfully I manually triggered the  secret rotation but I got a lambda timeout error (timeout was set to 30 secs) ..

```
START RequestId: e598024b-e5d3-4619-94b8-e919fcf1a621 Version: $LATEST
[ERROR]	2023-04-06T15:50:07.301Z	e598024b-e5d3-4619-94b8-e919fcf1a621	setSecret: Unable to log into database with previous, current, or pending secret of secret arn arn:aws:secretsmanager:us-east-1:035458030717:secret:nextflow-aurora-mysql-MasterSecret-08Z0bn
[ERROR] ValueError: Unable to log into database with previous, current, or pending secret of secret arn arn:aws:secretsmanager:us-east-1:035458030717:secret:nextflow-aurora-mysql-MasterSecret-08Z0bn
Traceback (most recent call last):
  File "/var/task/lambda_function.py", line 75, in lambda_handler
    set_secret(service_client, arn, token)
  File "/var/task/lambda_function.py", line 195, in set_secret
    raise ValueError("Unable to log into database with previous, current, or pending secret of secret arn %s" % arn)
END RequestId: e598024b-e5d3-4619-94b8-e919fcf1a621
REPORT RequestId: e598024b-e5d3-4619-94b8-e919fcf1a621	Duration: 40683.06 ms	Billed Duration: 40684 ms	Memory Size: 128 MB	Max Memory Used: 72 MB
```

I then manually increased the timeout to 1 minute, i'm not sure how you would set the timeout in cloudformation.  Then I manually re-triggered the secret rotation again.  This time the lambda failed to access the database.  The error..
```
START RequestId: e598024b-e5d3-4619-94b8-e919fcf1a621 Version: $LATEST
[ERROR]	2023-04-06T15:50:07.301Z	e598024b-e5d3-4619-94b8-e919fcf1a621	setSecret: Unable to log into database with previous, current, or pending secret of secret arn arn:aws:secretsmanager:us-east-1:035458030717:secret:nextflow-aurora-mysql-MasterSecret-08Z0bn
[ERROR] ValueError: Unable to log into database with previous, current, or pending secret of secret arn arn:aws:secretsmanager:us-east-1:035458030717:secret:nextflow-aurora-mysql-MasterSecret-08Z0bn
Traceback (most recent call last):
  File "/var/task/lambda_function.py", line 75, in lambda_handler
    set_secret(service_client, arn, token)
  File "/var/task/lambda_function.py", line 195, in set_secret
    raise ValueError("Unable to log into database with previous, current, or pending secret of secret arn %s" % arn)
END RequestId: e598024b-e5d3-4619-94b8-e919fcf1a621
REPORT RequestId: e598024b-e5d3-4619-94b8-e919fcf1a621	Duration: 40683.06 ms	Billed Duration: 40684 ms	Memory Size: 128 MB	Max Memory Used: 72 MB	
```

**Pull request closed without merging**
